UM/FlameProfiler.py

```diff
@@ -0,0 +1,187 @@
+    fill_children = _fillInProfileSpaces(start_time, end_time, raw_profile_calls)
+    return _ProfileCallNode("", 0, start_time, end_time, fill_children)
+
+##  Erase any profile data.
+def clearProfileData():
+    global clear_profile_requested
+    clear_profile_requested = True
+
+##  Start recording profile data.
+def startRecordingProfileData():
+    global record_profile_requested
+    record_profile_requested = True
+
+##  Stop recording profile data.
+def stopRecordingProfileData():
+    global stop_record_profile_requested
+    stop_record_profile_requested = True
+
+def _fillInProfileSpaces(start_time, end_time, profile_call_list):
+    result = []
+    time_counter = start_time
+    for profile_call in profile_call_list:
+        if secondsToMS(profile_call.getStartTime()) != secondsToMS(time_counter):
+            result.append(_ProfileCallNode("", 0, time_counter, profile_call.getStartTime(), []))
+        result.append(profile_call)
+        time_counter = profile_call.getEndTime()
+
+    if secondsToMS(time_counter) != secondsToMS(end_time):
+        result.append(_ProfileCallNode("", 0, time_counter, end_time, []))
+
+    return result
+
+def secondsToMS(value):
+    return math.floor(value *1000)
+
+##  Profile a block of code.
+#
+#   Use this context manager to wrap and profile a block of code.
+#   \param name \type{str} The name to use to identify this code in the profile report.
+@contextmanager
+def profileCall(name):
+    start_time = time.time()
+    child_accu_stack.append([])
+    yield
+    end_time = time.time()
+    call_stat = _ProfileCallNode(name, 0, start_time, end_time,
+                                 _fillInProfileSpaces(start_time, end_time, child_accu_stack.pop()))
+    child_accu_stack[-1].append(call_stat)
+
+##  Return whether we are recording profiling information.
+#
+#   \return \type{bool} True if we are recording.
+def isRecordingProfile():
+    global record_profile
+    return record_profile
+
+
+def updateProfileConfig():
+    global child_accu_stack
+    global record_profile
+
+    # We can only update the active profiling config when we are not deeply nested inside profiled calls.
+    if len(child_accu_stack) <= 1:
+        global clear_profile_requested
+        if clear_profile_requested:
+            clear_profile_requested = False
+            child_accu_stack = [[]]
+
+        global record_profile_requested
+        if record_profile_requested:
+            record_profile_requested = False
+            record_profile = True
+            Logger.log('d', 'Starting record record_profile_requested')
+
+        global stop_record_profile_requested
+        if stop_record_profile_requested:
+            stop_record_profile_requested = False
+            record_profile = False
+            Logger.log('d', 'Stopping record stop_record_profile_requested')
+
+##  Decorator which can be manually applied to methods to record profiling information.
+#
+#   \type{Callable}
+def profile(function):

```

Comment:
Can we use something less generic for this? This is going to interfere with the python line_profiler module which I use a lot to profile single functions.